Restructure documentation

phackstock opened the pull request:
closes #108.

This is my first iteration of the restructured docs. Still a little work in progress but the structure should be a lot better now.
For an easier overview I've written up the outline below.

## New structure
* **Index**
  * **Overview**
  * **Table of Contents**
  * **Acknowledgement**

* **Getting started**
  * **Package Overview** *short overview of the package*
  * **Installation** *how to install form pip or GH*
    * **Via Pip**
    * **From Source**
  * **Standard use case** *background on where nomenclature is typically used*
  * **Command line interface** *single example of the CLI with reference to the full section*
* **User Guide** *main part of the guide*
  * **Introduction**
  * **Directory structure** *structure of definitions and mappings*
  * **Codelist** *details about codelists*
    * **Variable** *special requirements for the variable codelist*
      * **Optional attributes**
      * **Variable naming conventions** *(not sure if this really belongs here)*
    * **Region** *special requirements for the region codelist*
    * **Generic** *some more detail about generic codelists*
    * **Tag** *details about using tags in codelists*
  * **Model mapping**
  * **Minimum working example** *all pieces together*
  * **Region processing** *some details about region processing (for now just partial region aggregation)*
    * **Partial region aggregation** *details of the logic of combining model native and aggregated results*
  * **CLI** *for now just autodocumented click documentation*
* **API**

danielhuppmann commented on doc/source/getting_started.rst:
```diff
@@ -1,116 +1,82 @@
-.. _getting-started:
+.. _getting_started:
```
This "Getting started" section seems to be pretty much useless. The package-overview is a repetition of the index page as well as the user-side-overview page.

And the section on "Standard Use Cases" will probably be overlooked by many casual users.

So I suggest to revert this page to a simple "Installation" page, and move the overview of standard use cases to the User Guide overview.

phackstock replied:
Renamed the section and integrated the "Standard Use Cases" into the user guide.

danielhuppmann commented on doc/source/getting_started.rst:
```diff
@@ -1,116 +1,82 @@
-  df = process(df, dsd, processor=rp)
 
-**Notes**
+Command line interface
```
Move this section to the CLI page.

phackstock replied:
Moved to the CLI page.

danielhuppmann commented on doc/source/index.rst:
```diff
@@ -37,26 +42,25 @@ Those two tasks are carried out by two classes:
+   information given in yaml model mapping files.
 
-A starting point on how to use these two classes with existing data templates and model mappings can be found in :ref:`getting-started`.
+Instructions on how to install and get started using nomenclature can be found in
```
Is it really necessary to add explicit references to these sections when the TOC is shown just below?

phackstock replied:
Removed the references.

danielhuppmann commented on doc/source/user_guide.rst:
```diff
@@ -0,0 +1,38 @@
+User Guide
+==========
+
+There are two main classes that the user interacts with when using the nomenclature
```
Start here with the main use cases: validation against a codelist and region-processing (the use case, not the class description).

phackstock replied:
Restructured it accordingly.

danielhuppmann commented on doc/source/user_guide.rst:
```diff
@@ -0,0 +1,38 @@
+Before that, however, the required directory structure for definitions and mappings as
+well as the file specifications for codelists and model mappings are discussed. 
+
+
+
+
```
Remove empty lines
```suggestion
```

danielhuppmann commented on doc/source/user_guide.rst:
```diff
@@ -0,0 +1,38 @@
+RegionProcessor and validated using DataStructureDefinition. Details are covered in
+:ref:`minimum_working_example`. 
+
+Before that, however, the required directory structure for definitions and mappings as
```
I do not think that many users will read this documentation like a novel, where you start with section 1 and then continue. This sentence can be removed without loss of information.

phackstock replied:
Fair point, removed the sentence.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,209 @@
+Codelist
+========
+
+A "codelist" (named in reference to code lists in `SDMX <https://sdmx.org/>`_) is a yaml
```
A codelist is a list of codes. In this package, the codelists are parsed from yaml files.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,209 @@
+  - ...
+
+In case of the "model" or "scenario" index dimension, such a simple list sufficient.
+More details on 
```
This sentence is not complete. However, I would omit this entire paragraph. Users will scroll through the page for the heading that they are interested in (variables, regions, ...).

phackstock replied:
Sounds good, removed it.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,209 @@
+example, if a codelist is located in a folder called ``model/`` it will be applied to
+the "model" index dimension. 
+
+A codelist must only contain a single list of allowed values. In the most simple case
```
What is the message here?
"A codelist must only contain a single list of allowed values."

phackstock replied:
The message was supposed to be that a codelist can only be a single list meaning something like this:

```yaml
allowed_regions:
  - region_a
  - region_b
  - ...
allowed_variables:
  - variable_a
  - ...
```

would not be allowed as this would be two lists (which are part of a mapping).
But you're right, the sentence probably does more harm than good and I think it can be removed without loss of information.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,209 @@
+--------
+
+An entry in a variable code list, **must be** a mapping (translated to python as a
+dictionary) maps the **name** of an allowed variable to, at minimum, a key value pair
```
..must be a mapping maps the name..

phackstock replied:
Corrected the sentence.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,209 @@
+Generic
+-------
+
+As mentioned in the beginning of this section, for IAMC dimensions other than 'region'
```
```suggestion
For IAMC dimensions other than 'region'
```

danielhuppmann commented on doc/source/user_guide/model-mapping.rst:
```diff
@@ -0,0 +1,75 @@
+.. _model_mapping:
+
+Model mapping
```
This should be merged with the Page "Region processing"

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,209 @@
+.. _codelist:
+
+Codelist
```
```suggestion
Validation using Codelists
```

danielhuppmann commented on doc/source/user_guide/region-processing.rst:
```diff
@@ -0,0 +1,42 @@
+Region processing
+=================
+
+Partial region aggregation
```
Why is this called "partial" region aggregation? It is a fully functional region aggregation according to three rules.

danielhuppmann reviewed (commented):
Thanks @phackstock for the first stab at re-organizing of the docs. I made a few suggestions to restructure further and streamline the content.

phackstock commented:
Addressed all your comments @danielhuppmann, 
I think it's starting to come together.
Apart from addressing your comments I also slightly restructured the sections about "Validation using Codelists" and "Region processing using model mappings". These sections are now structured very similarly, they contain two main parts:
* the format definition (for codelists and model mappings respectively)
* a short code example on how to use the corresponding classes 

There are currently two things that I'm not 100% happy with:
* In the Overview of the `index.rst` page and the first part of `user_guide.rst` the `RegionProcessor` and `DataStructureDefinition `classes are mentioned explicitly. Should I keep this or remove it from one/or both and keep the introductions more high level?
* The CLI part is a bit of a mix of hand-written and auto-generated content. Do you think that this is fine, should I go for a better mix of both or choose either completely auto-generated or hand-written. One worthwhile option would probably be to improve the docstrings for the cli functions and improve clarity this way.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+Codelist format specification
+-----------------------------
+
+In the most simple case of a codelist it contains a list of strings, e.g.:
```
```suggestion
In the most simple case, a codelist consists of a list of strings, e.g.:
```

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+The **unit** attribute is **required** and its value should be compatible with the
+`iam-units <https://github.com/iamconsortium/units>`_ package.
+
+The unit attribute can be:
```
This is not formatted as a list on RTD...

phackstock replied:
Added a second line break, now it's rendered correctly.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+While not strictly necessary a *description* attribute with a short description of the
+variable is encouraged. 
+
+In principle, *any* number of additional arbitrary named attributes are allowed. There
+are several attributes that affect the region-processing by the nomenclature package
+(see :ref:`region_aggregation_attributes`).
```
```suggestion
The yaml format allows *any* number of additional arbitrary named attributes. 
```

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+
+Optional attributes for region aggregation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
```suggestion
There  are several attributes that affect the region-processing by the nomenclature package.
See the Section :ref:`model_mapping` for more information.
```

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+       - Some Variable:
+         skip-region-aggregation: true
+
+  Setting *skip-region-aggregation* to ``true`` only skips the variable in question for
```
This sentence would be easier to understand if you cut it into two sentences.

phackstock replied:
Good point, split in two.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+                  method: max
+
+
+Guidelines and variable naming conventions
```
I suggest to move this "Guideline" section above the section "Optional attributes for region aggregation", because it is "more important" (everyone should be aware of these guidelines, users don't necessarily need to know about the region-aggregation attributes...

phackstock replied:
Switched the two.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+Region
+^^^^^^
+
+As is the case for the "variable" codelist, a region codelist must also follow a
+specific structure. 
+
```
This sentence does not add relevant information.
```suggestion
```

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+   - Hierachy 2:
+     - ...  
+
+Attributes of the **regions**, in the above example *some attribute* of *region 1* are
+entirely optional and have no influence on the region-processing.
+
+Nonetheless, they can be very useful, examples are: ISO2/3-codes
+(https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) or the list of countries
+included in a macro-region (i.e., a continent or large region).
```
Saying that something is unrelated to something else can be confusing if users wouldn't even have thought that they are related in the first place...
```suggestion
Useful examples of region attriutes are: ISO2/3-codes
(https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes)
or the list of countries included in a macro-region (i.e., a continent).
```

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+
+   - scenario 1
+   - scenario 2:
+     description: Something about scenario 1
```
This needs to be indented, right?

danielhuppmann commented on doc/source/user_guide/model-mapping.rst:
```diff
@@ -0,0 +1,149 @@
+      necessary to check the logs to find out if there were any differences. This is
+      intentional since some differences might be expected.
+
+The process of combining aggregated and model native results is dubbed
```
Is this name really used anywhere, other than in some tests? If not, I suggest removing this entirely.

phackstock replied:
Good point, it's not, let's leave it out then.

danielhuppmann commented on doc/source/user_guide/model-mapping.rst:
```diff
@@ -0,0 +1,149 @@
+The process of combining aggregated and model native results is dubbed
+'partial-region-aggregation'.
+
+The `region-aggregation` attribute (see :ref:`region_aggregation_attributes`) works with
```
This is already fully described in the list above.

phackstock replied:
Good point, I removed this and the previous paragraph as it was unnecessary information.

danielhuppmann commented on doc/source/user_guide/codelist.rst:
```diff
@@ -0,0 +1,244 @@
+
+In practice, ``DataStructureDefinition.validate`` is usually not called directly but
+rather as part of the :func:`process` function which combines validation and region
+processing. Further details on that are found in :ref:`minimum_working_example`.
```
```suggestion
processing. Refer to the :ref:`minimum_working_example` for details.
```

danielhuppmann commented on doc/source/user_guide/model-mapping.rst:
```diff
@@ -0,0 +1,149 @@
+Calling :meth:`RegionProcessor.apply` directly as in the example above is done mostly
+for illustrative purposes. 
+
+In practice, normally, both region processing *and* validation are desired and so the
+use of the :func:`process` function (details can be found under
+:ref:`minimum_working_example`) which combines both is recommended. Nonetheless, if only
+region processing is necessary the above code is appropriate.
```
```suggestion
In practice, normally, both region processing *and* validation are desired and so the
use of the :func:`process` function. Details can be found in the
:ref:`minimum_working_example`.
```

danielhuppmann commented on doc/source/user_guide/model-mapping.rst:
```diff
@@ -0,0 +1,149 @@
+Calling :meth:`RegionProcessor.apply` directly as in the example above is done mostly
+for illustrative purposes. 
+
+In practice, normally, both region processing *and* validation are desired and so the
```
```suggestion
In practice, both region processing *and* validation are desired and so the
```

danielhuppmann commented on doc/source/user_guide/model-mapping.rst:
```diff
@@ -0,0 +1,149 @@
+   processed_data = rp.apply(data)
+
+
+Calling :meth:`RegionProcessor.apply` directly as in the example above is done mostly
+for illustrative purposes. 
+
```
```suggestion
```

danielhuppmann reviewed (approved):
Thanks @phackstock. Agree, this is now in a much shape than before! I made a few more minor suggestions, then I think it's good to be merged.

Re your two open questions, I'll start a PR on a revised index page and package overview, and I think that the current documentation of the CLI is good enough for now.
